Use Grafana Agent for collecting metrics, logs, and traces

56quarters opened the pull request:
Use a DaemonSet of the Grafana Agent to collect metrics (using host filtering),
logs (mounting the node log directory), and traces (using a service for accepting
various thrift protocol traces).

This change configures the Prometheus instance to accept remote writes (from the
Grafana Agent) and not to scrape any pods (since this is handled by the agent).
It removes promtail (Loki client) and instead uses the promtail embedded in the
Grafana Agent.

Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

jdbaldry commented:
I've avoided Jsonnet formatting fixes in this PR but will happily contribute some linting and formatting if it is desirable. I wonder if we have a common Makefile for this since a number of projects have Jsonnet?

tomwilkie commented on production/sample/default/main.jsonnet:
```diff
@@ -45,7 +49,7 @@ prometheus + promtail + tns_mixin + {
 
   _images+:: {
-    grafana: 'grafana/grafana:6.6.1',
+    grafana: 'grafana/grafana:7.0.3',
```
Can you do this in grafana/jsonnet-libs instead?  And remove it from here.

buro9 replied:
Ah! This was my change in my local repo (before I handed over the laptop and lost the change). Thanks for the pointer 👍

tomwilkie commented:
> I've avoided Jsonnet formatting fixes in this PR but will happily contribute some linting and formatting if it is desirable. 

Yes please!

> I wonder if we have a common Makefile for this since a number of projects have Jsonnet?

Generally not found it to be worth the hassle; just copy and paste it from jsonnet-libs.

jdbaldry commented:
Recent changes to the prometheus-ksonnet library have changed the instance label on scraped jobs (from `<pod>` to `<pod>:<container>:<containerPortName>`) which breaks the pivot in Explore from Prometheus to Loki on the instance label. You can still pivot on the job label. 

To me, it feels like instance should match between Loki and Prometheus but the problem arises that a single container can have multiple instances of Prometheus metrics but only one of logs.

tomwilkie commented:
Can you work with @beorn7 and @slim-bean to come up with a solution? We recognised this would happen when we changed the prometheus relabelling rules, but didn't follow up.

I think the prometheus relabelling rules are fine now, Id say we should drop the instance label from loki (and use a pod label instead).

beorn7 commented:
Yeah, I think `pod` is the label to go with. Plus `container`, if Loki takes log streams per container. 

cluster/namespace/pod/container is the tuple I'd prefer, generally. It follows the drilldown trajectory.

jdbaldry commented:
Thanks @beorn7, that makes a lot of sense. I made sure to double check the labels in that tuple for parity between Promtail and Prometheus relabelling and realized that I needed to bump the vendored promtail too due to the container_name -> container changes in k8s 1.16. 

I'm pretty happy with this demo now, you can pivot on all the important labels as mentioned and now we get that sweet internal trace view.

56quarters committed:
Use Grafana Agent for collecting metrics, logs, and traces

Use a DaemonSet of the Grafana Agent to collect metrics (using host filtering),
logs (mounting the node log directory), and traces (using a service for accepting
thrift, zipkin, and oltp traces).

This change configures the Prometheus instance to accept remote writes (from the
Grafana Agent) and not to scrape any pods (since this is handled by the agent).
It removes promtail (Loki client) and instead uses the promtail embedded in the
Grafana Agent.

Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

56quarters commented:
NB: The motivation for this change is to have a simple, canonical example we can point to from documentation that uses a single Agent for the whole stack including exemplars. An exemplar of exemplars if you will :wink:

rfratto commented on create-k3d-cluster:
```diff
@@ -10,11 +10,17 @@ API_PORT=8555
 
-mkdir -p .volume
+mkdir -p $VOLUME
+if [ -f /etc/machine-id ]; then
+  cp -f /etc/machine-id $VOLUME/
+else
+  touch $VOLUME/machine-id
+fi
```
Should this be explained with a comment? It might not be obvious to readers why this is happening.

56quarters replied:
Good call :+1:

56quarters replied:
Fixed in c5d37f7dae886fb8b19c09218b7b7e607fba1224

rfratto commented on production/sample/default/main.jsonnet:
```diff
@@ -15,42 +14,122 @@ local tns_mixin = import 'tns-mixin/mixin.libsonnet';
+    gragent.withPortsMixin([
+      // Create container ports for the various ways that the agent can collect traces. Note that it's
+      // important to create the ports before creating the service for the agent since the method that
+      // creates the service will look at container ports to know which ports the service should have.
```
Oh, really? That's surprising, I would've thought that given the merge logic in Jsonnet it doesn't matter which you declare first.

56quarters replied:
I might be reading the Jsonnet wrong, but it seems like the `serviceFor` method here relies on the ports existing on the container already when creating the service: https://github.com/grafana/jsonnet-libs/blob/264a5c2078c5930af57fe2d107eff83ab63553af/ksonnet-util/util.libsonnet#L20

rfratto replied:
That's true, but I thought it would only be evaluated post-merge and not from top-down.

56quarters replied:
I'll try it out locally, changing up the order.

56quarters replied:
You're right, I'll remove the comment.

56quarters replied:
Fixed in e45543138433193c7ed21b5ff21807bd21c28c19

rfratto commented on production/sample/default/main.jsonnet:
```diff
@@ -15,42 +14,122 @@ local tns_mixin = import 'tns-mixin/mixin.libsonnet';
+      containerPort.new('thrift-binary', 6832) + containerPort.withProtocol('UDP'),
+      containerPort.new('thrift-http', 14268),
+      containerPort.new('thrift-grpc', 14250),
+      containerPort.new('zipkin', 9411),
```
Do you need a port for zipkin? It doesn't look like you're creating a zipkin receiver. Same might apply for other ports, since there's 4 receivers but 6 exposed ports here.

56quarters replied:
I'll doublecheck this.

56quarters replied:
Fixed in c5d37f7dae886fb8b19c09218b7b7e607fba1224

rfratto commented on production/sample/default/main.jsonnet:
```diff
@@ -15,42 +14,122 @@ local tns_mixin = import 'tns-mixin/mixin.libsonnet';
+      },
+      metrics+: {
+        global+: {
+          scrape_interval: '15s',
```
nit: should this be 60s to align with Grafana Cloud pricing?

56quarters replied:
Good call, I'll update

56quarters replied:
Fixed in c5d37f7dae886fb8b19c09218b7b7e607fba1224

rfratto reviewed (approved):
Nice :+1: I only gave this a high level review, so I don't know if the agent configs are "correct," but structurally this LGTM.

56quarters committed:
Code review fixes

Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

56quarters committed:
Code review fixes

Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>
